controllers/stardoginstance_controller.go

```diff
@@ -7,36 +7,293 @@ package controllers
+	r.Log.Info("adding Finalizers for the StardogInstance")
+	if !contains(rc.cfg.GetFinalizers(), stardogUsersFinalizer) {
+		if err := r.addFinalizer(rc, stardogUsersFinalizer); err != nil {
+			return err
+		}
+	}
+	if !contains(rc.cfg.GetFinalizers(), stardogRolesFinalizer) {
+		if err := r.addFinalizer(rc, stardogRolesFinalizer); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *StardogInstanceReconciler) addFinalizer(rc *ReconciliationContext, finalizer string) error {
+	r.Log.Info(fmt.Sprintf("Adding %s for the StardogInstance", finalizer))
+	controllerutil.AddFinalizer(rc.cfg, finalizer)
+	err := r.Update(rc.ctx, rc.cfg)
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("Cannot add %s for the StardogInstance", finalizer))
+		return err
+	}
+	return nil
+}
+
+func (r *StardogInstanceReconciler) deleteStardogInstance(rc *ReconciliationContext) error {
+	r.Log.Info("checking if StardogInstance is deletable")
+	stardogInstance := rc.cfg
+	if contains(stardogInstance.GetFinalizers(), stardogUsersFinalizer) {
+		if err := r.finalize(rc, stardogUsersFinalizer); err != nil {
+			return err
+		}
+		controllerutil.RemoveFinalizer(stardogInstance, stardogUsersFinalizer)
+		err := r.Update(rc.ctx, stardogInstance)
+		if err != nil {
+			return err
+		}
+	}
+	if contains(stardogInstance.GetFinalizers(), stardogRolesFinalizer) {
+		if err := r.finalize(rc, stardogRolesFinalizer); err != nil {
+			return err
+		}
 
-	// your logic here
+		controllerutil.RemoveFinalizer(stardogInstance, stardogRolesFinalizer)
+		err := r.Update(rc.ctx, stardogInstance)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *StardogInstanceReconciler) finalize(rc *ReconciliationContext, finalizer string) error {
+	username, password, err := r.getCredentials(rc.ctx, rc.cfg.Spec.AdminCredentials)
+	if err != nil {
+		return err
+	}
+
+	stardog := rest.NewWithBaseURIAndAuthentication(rc.cfg.Spec.ServerUrl, username, password)
+	if finalizer == stardogUsersFinalizer {
+		users, err := stardog.ListUsers(rc.ctx)
+		if err != nil {
+			r.Log.Error(err, "cannot list Stardog users", "namespace", rc.cfg.Namespace)
+			return err
+		}
+		if len(*users.Users) > 0 {
+			return fmt.Errorf("cannot delete StardogInstance, found %d users", len(*users.Users))
+		}
+	}
+	if finalizer == stardogRolesFinalizer {
+		roles, err := stardog.ListRoles(rc.ctx)
+		if err != nil {
+			r.Log.Error(err, "cannot list Stardog users", "namespace", rc.cfg.Namespace)
+			return err
+		}
+		if len(*roles.Roles) > 0 {
+			return fmt.Errorf("cannot delete StardogInstance, found %d roles", len(*roles.Roles))
+		}
+	}
+	return nil
+}
+
+func (r *StardogInstanceReconciler) validateSpecification(spec StardogInstanceSpec) error {
+	r.Log.Info("validating specification of StardogInstance CRD")
+	if len(spec.ServerUrl) == 0 {
+		return fmt.Errorf(".spec.ServerUrl is required")
+	}
+	if len(spec.AdminCredentials.SecretRef) == 0 {
```

Comment:
Compare with string equality
```suggestion
	if spec.AdminCredentials.SecretRef == "" {
```